terraform/jenkins/jenkins_controller.tf

```diff
@@ -0,0 +1,148 @@
+  virtual_machine_size         = "Standard_D1_v2"
+  virtual_machine_source_image = module.jenkins_controller_image.image_id
+
+  virtual_machine_custom_data = join("\n", ["#cloud-config", yamlencode({
+    users = [
+      for user in toset(["bmg", "flokli", "hrosten"]) : {
+        name                = user
+        sudo                = "ALL=(ALL) NOPASSWD:ALL"
+        ssh_authorized_keys = local.ssh_keys[user]
+      }
+    ]
+    write_files = [
+      # See corresponding EnvironmentFile= directives in services
+      {
+        content = "KEY_VAULT_NAME=${azurerm_key_vault.ssh_remote_build.name}\nSECRET_NAME=${azurerm_key_vault_secret.ssh_remote_build.name}",
+        "path"  = "/var/lib/fetch-build-ssh-key/env"
+      },
+      {
+        content = "KEY_VAULT_NAME=${azurerm_key_vault.binary_cache_signing_key.name}\nSECRET_NAME=${azurerm_key_vault_secret.binary_cache_signing_key.name}",
+        "path"  = "/var/lib/fetch-binary-cache-signing-key/env"
+      },
+      {
+        content = "AZURE_STORAGE_ACCOUNT_NAME=${azurerm_storage_account.binary_cache.name}",
+        "path"  = "/var/lib/rclone-http/env"
+      },
+      # Render /etc/nix/machines with terraform. In the future, we might want to
+      # autodiscover this, or better, have agents register with the controller,
+      # rather than having to recreate the VM whenever the list of builders is
+      # changed.
+      {
+        content = join("\n", [
+          for ip in toset(module.builder_vm[*].virtual_machine_private_ip_address) : "ssh://remote-build@${ip} x86_64-linux /etc/secrets/remote-build-ssh-key 10 10 kvm,big-parallel - -"
+        ]),
+        "path" = "/etc/nix/machines"
+      },
+      # Render /var/lib/builder-keyscan/scanlist, so known_hosts can be populated.
+      {
+        content = join("\n", toset(module.builder_vm[*].virtual_machine_private_ip_address))
+        "path"  = "/var/lib/builder-keyscan/scanlist"
+      }
+    ]
+  })])
+
+  allocate_public_ip = true
+  subnet_id          = azurerm_subnet.jenkins.id
+
+  # Attach disk to the VM
+  data_disks = [{
+    name            = azurerm_managed_disk.jenkins_controller_jenkins_state.name
+    managed_disk_id = azurerm_managed_disk.jenkins_controller_jenkins_state.id
+    lun             = "10"
+    # create_option = "Attach"
+    caching      = "None"
+    disk_size_gb = azurerm_managed_disk.jenkins_controller_jenkins_state.disk_size_gb
+  }]
+}
+
+resource "azurerm_network_interface_security_group_association" "jenkins_controller_vm" {
+  network_interface_id      = module.jenkins_controller_vm.virtual_machine_network_interface_id
+  network_security_group_id = azurerm_network_security_group.jenkins_controller_vm.id
+}
+
+resource "azurerm_network_security_group" "jenkins_controller_vm" {
+  name                = "jenkins-controller-vm"
+  resource_group_name = azurerm_resource_group.default.name
+  location            = azurerm_resource_group.default.location
+
+  security_rule {
```

Comment:
why do we need inbound ssh on Jenkins controller?

Reply:
So I could ssh in and debug the config. Also, until Jenkins is configured declaratively and uses SSO for login and RBAC, the admin password needs to be fetched from the logs/some file in the jenkins state.

Reply:
please close on production system, unless there is other use than debug/provisioning.

Reply:
Yes, there's a block in the futurework section about SSO and declarative pipeline config.
At least parts of this need to be addressed.